### ISSUE TITLE
Running a workflow with a customized genmodel results in errors in workspace

### ISSUE THREAD

**Issue by cwiscount** [eclipse/xtext-core#530]
This issue was introduced in 2.13. 
I have set up a small example at the following repository: https://github.com/cwiscount/mydsl-emf-gen

After running the GenerateMyDsl.mwe2 workflow, there will be errors in two files
* MyDslSemanticSequencer.java
* MyDslParsingTest.xtend

The issues are on the import statements that are referencing classes generated with the EcoreGenerator.  

I debugged this a bit to try to figure out what changed between 2.12 and 2.13.
The [following changes](https://github.com/eclipse/xtext-core/commit/624e4fa53bcb06115ebffa40b8568b36199c0b9b#diff-8917a219cc0d334b775bd8e223f8e2da) in TypeReference.xtend seem to have caused the change in behavior:

- `getQualifiedName(EClass, ResourceSet)`: 
  - 2.12: Used `getQualifiedInterfaceName()` in `GenClassImpl`: uses `getInterfacePackageName()` method which accounts for the suffix (`getInterfacePackageSuffix()`) 
  - 2.13: Changed to `getQualifiedPackageName()` in `GenPackageImpl`: Only uses base package
- `getQualifiedName(EPackage, ResourceSet)`:
  - 2.12: Used `getQualifiedPackageInterfaceName()` in `GenPackageImpl`: uses `getReflectionPackageName()` method which accounts for the suffix (`getMetaDataPackageSuffix()`) 
  - 2.13: Changed to `getQualifiedPackageName()` in `GenPackageImpl`: Only uses base package

In the example listed above, I have changed the Implementation, Interface, and Metadata Package Suffixes in the genmodel. The `getQualifiedPackageName()` method in GenPackageImpl does not appear to account for any of the suffixes that may be defined in the genmodel, which is why the code is generating with errors. 

If you require any more feedback, or have any questions, please don't hesitate to ask.

**Comment by cdietrich** [comment]
hmmmm thats really bad. we changed this cause there are gazillions of people out there that dont use naming conventions

**Comment by cdietrich** [comment]
So I wonder is simply using the two methods works

**Comment by cwiscount** [comment]
Do you mean using the methods from the 2.12 release in the place where they were updated to getQualifiedPackageName()? 

I was wondering that as well. If there is no suffix (i.e. it's empty), it should just return the package name returned by getQualifiedPackageName(). (I'm looking at addPackageSuffix in GenBaseImpl)

**Comment by cdietrich** [comment]
No I mean pick the right ones from the tons there are in Genpackage/genClass. The question is which are the right ones

**Comment by cwiscount** [comment]
Ah ok. 
What exactly was broken with the getQualifiedInterfaceName() and getQualifiedPackageInterfaceName() methods?

**Comment by cdietrich** [comment]
we need "package" and "type" name separately

**Comment by cdietrich** [comment]
=> i wonder if 

```
private static def QualifiedClassName getQualifiedName(EClass clazz, ResourceSet resourceSet) {
		if (clazz.EPackage.nsURI == 'http://www.eclipse.org/2008/Xtext') {
			new QualifiedClassName('org.eclipse.xtext', clazz.name)
		} else if (clazz.EPackage.nsURI == 'http://www.eclipse.org/emf/2002/Ecore') {
			if (clazz.instanceTypeName !== null) {
				val itn = clazz.instanceTypeName;
				new QualifiedClassName(itn.substring(0, itn.lastIndexOf('.')),
					itn.substring(itn.lastIndexOf(".") + 1).replace("$", "."))
			} else {
				new QualifiedClassName('org.eclipse.emf.ecore', clazz.name)
			}
		} else {
			new QualifiedClassName(GenModelUtil2.getGenClass(clazz, resourceSet).genPackage.reflectionPackageName,
				GenModelUtil2.getGenClass(clazz, resourceSet).interfaceName)
		}
	}

	private static def QualifiedClassName getQualifiedName(EPackage epackage, ResourceSet resourceSet) {
		new QualifiedClassName(GenModelUtil2.getGenPackage(epackage, resourceSet).reflectionPackageName,
			GenModelUtil2.getGenPackage(epackage, resourceSet).packageInterfaceName)
	}
``` is fine or if `packageInterfaceName` and `interfaceName` are pot. wrong as well.

**Comment by cwiscount** [comment]
I'm not sure if we can rely solely on getReflectionPackageName. What if we set the metadata and interface packages to different values? 

I wonder if getQualifiedName(EClass, ResourceSet) should be using `GenModelUtil2.getGenClass(clazz, resourceSet).genPackage.interfacePackageName` instead?

**Comment by cdietrich** [comment]
i have no idea what in EMF can be done to change the names around. do you have ideas whats possible?

so  maybe `getGenModel().isSuppressEMFMetaData() ? getQualifiedPackageClassName() : getReflectionPackageName()`  but what for the classes?

**Comment by cdietrich** [comment]
but yes interfacePackageName looks good

**Comment by cwiscount** [comment]
Here's a screen shot of the configs that deal with naming (I think this is all of them...I'm by no means an EMF expert) 

![ecoregenmodelconfig](https://user-images.githubusercontent.com/33351210/32511584-34e5f8dc-c3c2-11e7-8d3d-3775c42b81b3.PNG)

So as far as the EMF Package Suffixes go, it looks like the following can be changed: 
- Implementation
- Interface
- Metadata
- Presentation
- Provider
- Tests
- Utility

Then, you are also able to update the Base Package and Ecore Package. 
And the prefix seems to be used for generating the Factory, Package, AdapterFactory, and Switch classes. 

Here's another screenshot of what is generated with the config that I attached above: 
![ecoregen](https://user-images.githubusercontent.com/33351210/32511830-065e42fc-c3c3-11e7-9a19-67f11d049783.PNG)

**Comment by cdietrich** [comment]
i think we use interface and metadata only. i tought they where the same so far

**Comment by cwiscount** [comment]
Yeah, I'm pretty sure you only use interface and metadata as well. I'm not sure how often those two packages are different in practice, but the option is there, so I'm sure somebody has the two of them configured differently. In most, if not all, of our projects, we have them configured to the same package.

**Comment by cdietrich** [comment]
can you test if my fix solves your problem?

**Comment by cwiscount** [comment]
Sure, but how would I go about doing that? Do I need a dev environment set up, or do you have a build that I can pull from? I'm not familiar with your development procedures at this point.

**Comment by cdietrich** [comment]
can you test with e.g. maven/gradle? or eclipse only

**Comment by cdietrich** [comment]
you could try to setup 

http://services.typefox.io/open-source/jenkins/job/xtext-umbrella/job/master/419/artifact/build/p2-repository/
+
http://services.typefox.io/open-source/jenkins/job/xtext-core/job/cd_issue530/lastSuccessfulBuild/artifact/build/p2-repository/

**Comment by cwiscount** [comment]
Just tested it out in my eclipse and the import issue is corrected. I also set the packages to all be different and it worked as expected, so the fix looks good. Thank you! 

One side question, should the interface package be automatically added to the exported packages in the manifest of the core project? When I regenerated, I had errors in my tests plugin, but it was because the interfaces were not visible to that project. I added the export and all was well again. I'm wondering if there is a similar issue there where it is only adding the base package to the exports?

**Comment by cdietrich** [comment]
since you generate the package manually:no. it is up to EMF/you

**Comment by cdietrich** [comment]
thx for testing

**Comment by cdietrich** [comment]
Fixed in Xtext 2.14
